nexus/src/app/sagas/mod.rs

```diff
@@ -116,3 +116,129 @@ pub(super) async fn saga_generate_uuid<UserType: SagaType>(
 ) -> Result<Uuid, ActionError> {
     Ok(Uuid::new_v4())
 }
+
+macro_rules! __stringify_ident {
+    ($i:ident) => {
+        stringify!($i)
+    };
+}
+
+macro_rules! __emit_action {
+    ($node:ident, $output:literal) => {
+        paste::paste! {
+            #[allow(dead_code)]
+            fn [<$node:lower _action>]() -> ::steno::Node {
+                ::steno::Node::action(
+                    $output,
+                    crate::app::sagas::__stringify_ident!([<$node:camel>]),
+                    $node.as_ref(),
+                )
+            }
+        }
+    };
+}
+
+macro_rules! __action_name {
+    ($saga:ident, $node:ident) => {
+        paste::paste! {
+            concat!(
+                stringify!($saga),
+                ".",
+                crate::app::sagas::__stringify_ident!([<$node:lower>]),
+            )
+        }
+    };
+}
+
+/// A macro intended to reduce boilerplate when writing saga actions.
+///
+/// For this input:
+///
+/// ```ignore
+/// declare_saga_actions! {
+///     my_saga;
+///     SAGA_NODE1 -> "output1" {
+///         + do1
+///         - undo1
+///     }
+///     SAGA_NODE2 -> "output2" {
+///         + do2
+///     }
+/// }
+/// ```
+///
+/// We generate the following:
+/// - For `SAGA_NODE1`:
+///     - A `NexusAction` labeled "my_saga.saga_node1" (containing "do1" and "undo1").
+///     - `fn saga_node1_action() -> steno::Node` referencing this node, with an
+///     output named "output1".
+/// - For `SAGA_NODE2`:
+///     - A `NexusAction` labeled "my_saga.saga_node2" (containing "do2").
+///     - `fn saga_node2_action() -> steno::Node` referencing this node, with an
+///     output named "output2".
+/// - For `my_saga`:
+///     - `fn my_saga_register_actions(...)`, which can be called to implement
+///     `NexusSaga::register_actions`.
+macro_rules! declare_saga_actions {
```

Comment:
Do we expect to make this a public macro (`macro_export`) in the future, maybe for other parts of the code that use sagas?

Reply:
Good question! Right now, I expected to keep this "Nexus-internal", since my short-term goal was to just make it easier to add new sagas.

However, if we like the structure of this macro, I could see us making it public, or moving it into the `steno` repo itself.